### PR TITLE
Allow for throwing exceptions via `calling`

### PR DESCRIPTION
**Pull request by philomates**
In midje you can do 
```clojure
(fact
  (some-fn) => (throws Exception)
  (provided (some-fn) =throws=> (Exception.)))
```
which isn't currently possible in `mockfn`

This PR extends mockfn to allow for this behavior, and more, by distinguishing between mocking with a base value, or mocking with a function that should be called.

```clojure
(testing "providing - cause function to throw exception"
  (mfn/providing [(one-fn :argument-1) (mck/as-fn #(throw (ex-info "one-fn failedl" {:arg %})))
                  (one-fn :argument-2) :result-2]
    (is (thrown? ExceptionInfo (one-fn :argument-1)))
    (is (= :result-2 (one-fn :argument-2)))))
```

NOTE:
I haven't written very many tests because I want to demonstrate the idea before making it ready to merge. Once I get some initial comments I'll finish up the PR.

### CODE REVIEW

File: src/mockfn/mock.clj [hunk 1]
```diff
@@ -2,6 +2,12 @@
   (:require [mockfn.matchers :as matchers]
             [mockfn.parser]))
 
+(defrecord AsFunc [func])
+
+(defn as-fn
```
Comment (philomates) [inline src/mockfn/mock.clj]: I'm wondering if this should live in `mockfn.macros` to avoid having to import another namespace.

Comment (philomates) [reply src/mockfn/mock.clj]: Would something like `calling` or `calls` be a better name than `as-fn`?

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,10 @@ All notable changes to this project are documented in this file.
 
 ## Releases
 
+### 0.4.0 - 2019-01-07
+- Allow mocks to not only return base values, but invoke function calls.
+  This enables mocked functions to throw exceptions.
```
Comment (pmatiello) [inline CHANGELOG.md]: This can go in 0.3.0. We haven't made a release it.

File: project.clj [hunk 3]
```diff
@@ -1,6 +1,6 @@
-(defproject mockfn "0.3.0"
+(defproject mockfn "0.4.0"
```
Comment (pmatiello) [inline project.clj]: Same as above: 0.3.0.

File: src/mockfn/mock.clj (hunk 1 above)
Comment (pmatiello) [reply src/mockfn/mock.clj]: I like both ideas: moving it to `mockfn.macros` and renaming it to `calling`.

**Pull request merged**